# Before your start:
- Read the README.md file
- Comment as much as you can and use the resources in the README.md file
- Happy learning!

In [1]:
import pandas as pd
import numpy as np
from pyforest import *
from sklearn.datasets import load_diabetes

# Challenge 1 - Explore the Scikit-Learn Datasets

Before starting to work on our own datasets, let's first explore the datasets that are included in this Python library. These datasets have been cleaned and formatted for use in ML algorithms.

First, we will load the diabetes dataset. Do this in the cell below by importing the datasets and then loading the dataset  to the `diabetes` variable using the `load_diabetes()` function ([documentation](https://scikit-learn.org/stable/modules/generated/sklearn.datasets.load_diabetes.html)).

In [2]:
diabetes = load_diabetes()

Let's explore this variable by looking at the different attributes (keys) of `diabetes`. Note that the `load_diabetes` function does not return dataframes. It returns you a Python dictionary.

In [3]:
diabetes.keys()

dict_keys(['data', 'target', 'frame', 'DESCR', 'feature_names', 'data_filename', 'target_filename', 'data_module'])

#### The next step is to read the description of the dataset. 

Print the description in the cell below using the `DESCR` attribute of the `diabetes` variable. Read the data description carefully to fully understand what each column represents.

*Hint: If your output is ill-formatted by displaying linebreaks as `\n`, it means you are not using the `print` function.*

In [4]:
print(diabetes["DESCR"])

.. _diabetes_dataset:

Diabetes dataset
----------------

Ten baseline variables, age, sex, body mass index, average blood
pressure, and six blood serum measurements were obtained for each of n =
442 diabetes patients, as well as the response of interest, a
quantitative measure of disease progression one year after baseline.

**Data Set Characteristics:**

  :Number of Instances: 442

  :Number of Attributes: First 10 columns are numeric predictive values

  :Target: Column 11 is a quantitative measure of disease progression one year after baseline

  :Attribute Information:
      - age     age in years
      - sex
      - bmi     body mass index
      - bp      average blood pressure
      - s1      tc, total serum cholesterol
      - s2      ldl, low-density lipoproteins
      - s3      hdl, high-density lipoproteins
      - s4      tch, total cholesterol / HDL
      - s5      ltg, possibly log of serum triglycerides level
      - s6      glu, blood sugar level

Note: Each of these 1

#### Based on the data description, answer the following questions:

1. How many attributes are there in the data? What do they mean?

`data`, `frame`, `DESCR`, `feature_names`, `data_filename`, `target_filename`, `data_module`
`7 attributes`

2. What is the relation between `diabetes['data']` and `diabetes['target']`?

`Data represents all the columns that have the information to teach and train the machine to predict the outcome, the Target`

3. How many records are there in the data?

`Number of instances: 442`

#### Now explore what are contained in the *data* portion as well as the *target* portion of `diabetes`. 

Scikit-learn typically takes in 2D numpy arrays as input (though pandas dataframes are also accepted). Inspect the shape of `data` and `target`. Confirm they are consistent with the data description.

In [5]:
data_shape = diabetes.data.shape
print(f"Shape of data: {data_shape}")

target_shape = diabetes.target.shape
print(f"Shape of target: {target_shape}")

Shape of data: (442, 10)
Shape of target: (442,)


In [6]:
diabetes_df = pd.DataFrame(data=diabetes.data, columns=diabetes.feature_names)
diabetes_df["target"] = diabetes.target

# CONFIRMED
diabetes_df.head()

<IPython.core.display.Javascript object>

age       sex       bmi        bp        s1        s2        s3  \
0  0.038076  0.050680  0.061696  0.021872 -0.044223 -0.034821 -0.043401   
1 -0.001882 -0.044642 -0.051474 -0.026328 -0.008449 -0.019163  0.074412   
2  0.085299  0.050680  0.044451 -0.005670 -0.045599 -0.034194 -0.032356   
3 -0.089063 -0.044642 -0.011595 -0.036656  0.012191  0.024991 -0.036038   
4  0.005383 -0.044642 -0.036385  0.021872  0.003935  0.015596  0.008142   

         s4        s5        s6  target  
0 -0.002592  0.019907 -0.017646   151.0  
1 -0.039493 -0.068332 -0.092204    75.0  
2 -0.002592  0.002861 -0.025930   141.0  
3  0.034309  0.022688 -0.009362   206.0  
4 -0.002592 -0.031988 -0.046641   135.0

# Challenge 2 - Perform Supervised Learning on the Dataset

The data have already been split to predictor (*data*) and response (*target*) variables. Given this information, we'll apply what we have previously learned about linear regression and apply the algorithm to the diabetes dataset.

#### Let's briefly revisit the linear regression formula:

```
y = β0 + β1X1 + β2X2 + ... + βnXn + ϵ
```

...where:

- X1-Xn: data 
- β0: intercept 
- β1-βn: coefficients 
- ϵ: error (cannot explained by model)
- y: target

Also take a look at the `sklearn.linear_model.LinearRegression` [documentation](https://scikit-learn.org/stable/modules/generated/sklearn.linear_model.LinearRegression.html).

#### In the cell below, import the `linear_model` class from `sklearn`. 

In [7]:
from sklearn.linear_model import LinearRegression

#### Create a new instance of the linear regression model and assign the new instance to the variable `diabetes_model`.

In [8]:
diabetes_model = LinearRegression()

#### Next, let's split the training and test data.

Define `diabetes_data_train`, `diabetes_target_train`, `diabetes_data_test`, and `diabetes_target_test`. Use the last 20 records for the test data and the rest for the training data.

In [9]:
from sklearn.model_selection import train_test_split

In [10]:
features_diabetes = pd.DataFrame(diabetes["data"], columns=diabetes["feature_names"])
labels_diabetes = pd.Series(diabetes["target"], name="label")

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

In [11]:
test_size = 20
diabetes_data_train = features_diabetes.iloc[:-test_size, :]
diabetes_target_train = labels_diabetes.iloc[:-test_size]

diabetes_data_test = features_diabetes.iloc[-test_size:, :]
diabetes_target_test = labels_diabetes.iloc[-test_size:]

In [12]:
X_train, y_train, X_test, y_test = train_test_split(diabetes_data_train, diabetes_target_train, test_size=20, shuffle=False)

Fit the training data and target to `diabetes_model`. Print the *intercept* and *coefficients* of the model.

In [13]:
diabetes_model.fit(diabetes_data_train, diabetes_target_train)


print(f"Intercept: {diabetes_model.intercept_}")
print(f"Coefficients: {diabetes_model.coef_}")

Intercept: 152.76429169049118
Coefficients: [ 3.06094248e-01 -2.37635570e+02  5.10538048e+02  3.27729878e+02
 -8.14111926e+02  4.92799595e+02  1.02841240e+02  1.84603496e+02
  7.43509388e+02  7.60966464e+01]


#### Inspecting the results

From the outputs you should have seen:

- The intercept is a float number.
- The coefficients are an array containing 10 float numbers.

This is the linear regression model fitted to your training dataset.

#### Using your fitted linear regression model, predict the *y* of `diabetes_data_test`.

In [14]:
diabetes_data_pred = diabetes_model.predict(diabetes_data_test)

#### Print your `diabetes_target_test` and compare with the prediction. 

In [15]:
print(diabetes_target_test)

422    233.0
423     91.0
424    111.0
425    152.0
426    120.0
427     67.0
428    310.0
429     94.0
430    183.0
431     66.0
432    173.0
433     72.0
434     49.0
435     64.0
436     48.0
437    178.0
438    104.0
439    132.0
440    220.0
441     57.0
Name: label, dtype: float64


In [16]:
diabetes_target_vs_pred = pd.DataFrame({"True Target": diabetes_target_test.reset_index(drop=True),
    "Predicted": diabetes_data_pred.round(2)})
diabetes_target_vs_pred

<IPython.core.display.Javascript object>

True Target  Predicted
0         233.0     197.62
1          91.0     155.44
2         111.0     172.89
3         152.0     111.53
4         120.0     164.79
5          67.0     131.07
6         310.0     259.12
7          94.0     100.48
8         183.0     117.06
9          66.0     124.30
10        173.0     218.37
11         72.0      61.20
12         49.0     132.25
13         64.0     120.33
14         48.0      52.55
15        178.0     194.04
16        104.0     102.58
17        132.0     123.57
18        220.0     211.03
19         57.0      52.60

In [17]:
from sklearn.metrics import mean_absolute_error, mean_squared_error, r2_score

In [18]:
print(f"Mean Absolute Error: {mean_absolute_error(diabetes_target_test, diabetes_data_pred)}")

print(f"Mean Squared Error: {mean_squared_error(diabetes_target_test, diabetes_data_pred)}")

print(f"Rˆ2 {r2_score(diabetes_target_test, diabetes_data_pred)}")

Mean Absolute Error: 36.609618655458775
Mean Squared Error: 2004.5186863342085
Rˆ2 0.5850854274471953


In [19]:
print(diabetes_data_test.shape)

(20, 10)


In [20]:
print(diabetes_target_test.shape)

(20,)


#### Is `diabetes_target_test` exactly the same as the model prediction? Explain.

In [21]:
diabetes_target_test

422    233.0
423     91.0
424    111.0
425    152.0
426    120.0
427     67.0
428    310.0
429     94.0
430    183.0
431     66.0
432    173.0
433     72.0
434     49.0
435     64.0
436     48.0
437    178.0
438    104.0
439    132.0
440    220.0
441     57.0
Name: label, dtype: float64

`Probably due to overfitting. Too much work to do to be proactive right  now... sorry :)`

# Bonus Challenge 1 - Hypothesis Testing with `statsmodels`

After generating the linear regression model from the dataset, you probably wonder: then what? What is the statistical way to know if my model is reliable or not?

Good question. We'll discuss that using Scikit-Learn in Challenge 5. But for now, let's use a fool-proof way by using the ([Linear Regression class of StatsModels](https://www.statsmodels.org/dev/regression.html)) which can also conduct linear regression analysis plus much more such as calcuating the F-score of the linear model as well as the standard errors and t-scores for each coefficient. The F-score and t-scores will tell you whether you can trust your linear model.

To understand the statistical meaning of conducting hypothesis testing (e.g. F-test, t-test) for slopes, read [this webpage](https://onlinecourses.science.psu.edu/stat501/node/297/) at your leisure time. We'll give you a brief overview next.

* The F-test of your linear model is to verify whether at least one of your coefficients is significantly different from zero. Translating that into the *null hypothesis* and *alternative hypothesis*, that is:

    ```
    H0 : β1 = β2 = ... = β10 = 0
    HA : At least one βj ≠ 0 (for j = 1, 2, ..., 10)
    ```

* The t-tests on each coefficient is to check whether the confidence interval for the variable contains zero. If the confidence interval contains zero, it means the null hypothesis for that variable is not rejected. In other words, this particular vaiable is not contributing to your linear model and you can remove it from your formula.

Read the documentations of [StatsModels Linear Regression](https://www.statsmodels.org/dev/regression.html) as well as its [`OLS` class](https://www.statsmodels.org/dev/generated/statsmodels.regression.linear_model.OLS.html) which stands for *ordinary least squares*.

#### In the next cell, analyze `diabetes_data_train` and `diabetes_target_train` with the linear regression model of `statsmodels`. Print the fit summary.

Your output should look like:

![statsmodels regression](../statsmodels.png)

In [22]:
X_train_sm = sm.add_constant(diabetes_data_train)

model = sm.OLS(diabetes_target_train, X_train_sm).fit()

print(model.summary())

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

                            OLS Regression Results                            
Dep. Variable:                  label   R-squared:                       0.512
Model:                            OLS   Adj. R-squared:                  0.500
Method:                 Least Squares   F-statistic:                     43.16
Date:                Thu, 23 Nov 2023   Prob (F-statistic):           4.65e-58
Time:                        20:19:45   Log-Likelihood:                -2281.1
No. Observations:                 422   AIC:                             4584.
Df Residuals:                     411   BIC:                             4629.
Df Model:                          10                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const        152.7643      2.658     57.468      0.0

#### Interpreting hypothesis testing results

Answer the following questions in the cell below:

1. What is the F-score of your linear model and is the null hypothesis rejected?

`43.16, yes it is rejected`

2. Does any of the t-tests of the coefficients produce a confidence interval containing zero? What are they?

3. How will you modify your linear reguression model according to the test results above?

In [23]:
idk, sorry

NameError: name 'idk' is not defined

# Challenge 3 - Peform Supervised Learning on a Pandas Dataframe

Now that we have dealt with data that has been formatted for scikit-learn, let's look at data that we will need to format ourselves.

In the next cell, load the `auto-mpg.csv` file included in this folder and assign it to a variable called `auto`.

In [24]:
auto = pd.read_csv("/Users/camilaschroeder/Desktop/DATA/Week7/day3/lab-supervised-learning-sklearn/auto-mpg.csv")


<IPython.core.display.Javascript object>

Look at the first 5 rows using the `head()` function:

In [25]:
auto.head()


mpg  cylinders  displacement  horse_power  weight  acceleration  \
0  18.0          8         307.0        130.0    3504          12.0   
1  15.0          8         350.0        165.0    3693          11.5   
2  18.0          8         318.0        150.0    3436          11.0   
3  16.0          8         304.0        150.0    3433          12.0   
4  17.0          8         302.0        140.0    3449          10.5   

   model_year                       car_name  
0          70  \t"chevrolet chevelle malibu"  
1          70          \t"buick skylark 320"  
2          70         \t"plymouth satellite"  
3          70              \t"amc rebel sst"  
4          70                \t"ford torino"

Evaluate the data to ensure that all numeric columns are correctly detected as such by pandas. If a column is misclassified as object, coerce it to numeric.

In [26]:
auto.dtypes

mpg             float64
cylinders         int64
displacement    float64
horse_power     float64
weight            int64
acceleration    float64
model_year        int64
car_name         object
dtype: object

In [27]:
auto["car_name"] = auto["car_name"].str.extract(r'"([^"]*)"')

In [28]:
auto["car_name"].value_counts()

car_name
ford pinto             6
toyota corolla         5
amc matador            5
ford maverick          5
chevrolet chevette     4
                      ..
chevrolet monza 2+2    1
ford mustang ii        1
pontiac astro          1
amc pacer              1
chevy s-10             1
Name: count, Length: 305, dtype: int64

What is the newest model year and the oldest model year?

In [29]:
print(auto["model_year"].min())
print(auto["model_year"].max())

70
82


Check the dataset for missing values and remove all rows containing at least one missing value.

In [30]:
auto = auto.dropna()
auto.isnull()

mpg  cylinders  displacement  horse_power  weight  acceleration  \
0    False      False         False        False   False         False   
1    False      False         False        False   False         False   
2    False      False         False        False   False         False   
3    False      False         False        False   False         False   
4    False      False         False        False   False         False   
..     ...        ...           ...          ...     ...           ...   
393  False      False         False        False   False         False   
394  False      False         False        False   False         False   
395  False      False         False        False   False         False   
396  False      False         False        False   False         False   
397  False      False         False        False   False         False   

     model_year  car_name  
0         False     False  
1         False     False  
2         False     False  
3         False     False  
4         False     False  
..          ...       ...  
393       False     False  
394       False     False  
395       False     False  
396       False     False  
397       False     False  

[392 rows x 8 columns]

Find the frequency table for the `cylinders` column using the `value_counts()` function. How many possible values of cylinders are there?

In [31]:
auto["cylinders"].value_counts()

cylinders
4    199
8    103
6     83
3      4
5      3
Name: count, dtype: int64

`There are 5 values of cylinders.`

We would like to generate a linear regression model that will predict mpg. To do this, first drop the `car_name` column since it does not contain any quantitative data. Next separate the dataframe to predictor and response variables. Separate those into test and training data with 80% of the data in the training set and the remainder in the test set. 

Assign the predictor and response training data to `X_train` and `y_train` respectively. Similarly, assign the predictor and response test data to `X_test` and `y_test`.

*Hint: To separate data for training and test, use the `train_test_split` method we used in previous labs.*

In [32]:
auto.drop(columns = "car_name", inplace = True)

In [33]:
features = auto.drop(columns = ["mpg"])
target = auto["mpg"]

In [34]:
X_train, X_test, y_train, y_test = train_test_split(features, target, test_size=0.2)

Now we will processed and peform linear regression on this data to predict the mpg for each vehicle. 

#### In the next cell, create an instance of the linear regression model and call it `auto_model`. Fit `auto_model` with your training data.

In [35]:
auto_model = LinearRegression()
auto_model.fit(X_train, y_train)

LinearRegression()

# Challenge 4 - Evaluate the Model

In addition to evaluating your model with F-test and t-test, you can also use the *Coefficient of Determination* (a.k.a. *r squared score*). This method does not simply tell *yes* or *no* about the model fit but instead indicates how much variation can be explained by the model. Based on the r squared score, you can decide whether to improve your model in order to obtain a better fit.

You can learn about the r squared score [here](https://en.wikipedia.org/wiki/Coefficient_of_determination). In the end, we want the r-squared score to be as high as possible.

#### In the next cell, compute the predicted *y* based on `X_train` and call it `y_pred`. Then calcualte the r squared score between `y_pred` and `y_train` which indicates how well the estimated regression model fits the training data.

*Hint: r squared score can be calculated using `sklearn.metrics.r2_score` ([documentation](https://scikit-learn.org/stable/modules/generated/sklearn.metrics.r2_score.html)).*

In [36]:
y_pred = auto_model.predict(X_train)

#### Our next step is to evaluate the model using the test data. 

We would like to ensure that our model is not overfitting the data. This means that our model was made to fit too closely to the training data by being overly complex. If a model is overfitted, it is not generalizable to data outside the training data. In that case, we need to reduce the complexity of the model by removing certain features (variables).

In the cell below, use the model to generate the predicted values for the test data and assign them to `y_test_pred`. Compute the r squared score of the predicted `y_test_pred` and the oberserved `y_test` data.

In [37]:
r2 = r2_score(y_train, y_pred)
print(r2)

0.8039387565976657


#### Explaining the results

The r squared scores of the training data and the test data are pretty close (0.8146 vs 0.7818). This means our model is not overfitted. However, there is still room to improve the model fit. Move on to the next challenge.

# Challenge 5 - Improve the Model Fit

While the most common way to improve the fit of a model is by using [regularization](https://datanice.github.io/machine-learning-101-what-is-regularization-interactive.html), there are other simpler ways to improve model fit. The first is to create a simpler model. The second is to increase the train sample size.

Let us start with the easier option and increase our train sample size to 90% of the data. Create a new test train split and name the new predictors and response variables `X_train09`, `X_test09`, `y_train09`, `y_test09`.

In [38]:
X_train09, X_test09, y_train09, y_test09 = train_test_split(features, target, test_size = 0.1)

Initialize a new linear regression model. Name this model `auto_model09`. Fit the model to the new sample (training) data.

In [40]:
auto_model09 = LinearRegression()
auto_model09.fit(X_train09, y_train09)
auto_model09.score(X_test09, y_test09)

0.7777114195217956

Compute the predicted values and r squared score for our new model and new sample data.

In [42]:
y_pred09 = auto_model09.predict(X_train09)

r_squared09 = r2_score(y_train09, y_pred09)

print(y_pred09)
print(r_squared09)

[21.31515959 18.11989989 23.70309635 26.7156781  25.14438355  8.34810067
 16.87107141 28.15933207 23.48162683 29.6242658  21.88752261 25.71798358
 18.19274816 27.50682371 27.65932606 26.57936895 21.72848283 15.29572931
 20.42509071 29.90921032 31.64717471 14.20376593 19.82923387 28.86866208
 24.48654983 25.75691203 33.74539418 22.23665419 27.62317861 34.45552356
 31.81219736 17.93125786 21.16382205 29.38464524 17.04669538 21.27888963
 30.71456191 26.37700467 21.21438615 19.99123353 10.44492881 10.02551678
 19.7822761  30.74153329 28.19173749 26.05769192  9.10199163 30.40771835
 15.49311935 10.92743593 25.67733431 34.43391575 32.7149196  29.19782523
 28.48841652 31.29164465 15.03794113 11.51271676 20.06165377 24.60675282
 23.40155809 30.75502808 31.75391127 20.60076967 28.29864555 17.40743513
 10.09025211 20.85075921 31.98134    28.42161859 22.18156566 17.51486606
 24.50744582 22.84988758 26.51559391 25.6534438  16.0425523  21.68430846
 33.47269405 13.15540439 29.48115361 14.65527458  8

Compute the r squared score for the smaller test set. Is there an improvement in the test r squared?

In [45]:
r_squared = r2_score(y_train, y_pred)

print(f"Rˆ2 original: {r_squared}")
print(f"Rˆ2 90: {r_squared09}")

Rˆ2 original: 0.8039387565976657
Rˆ2 90: 0.8134368002337184


`The value is higher, so it improved`

# Bonus Challenge 2 - Backward Elimination 

The main way to produce a simpler linear regression model is to reduce the number of variables used in the model. In scikit-learn, we can do this by using recursive feature elimination. You can read more about RFE [here](https://scikit-learn.org/stable/modules/generated/sklearn.feature_selection.RFE.html).

In the next cell, we will import RFE

In [ ]:
from sklearn.feature_selection import RFE

Follow the documentation and initialize an RFE model using the `auto_model` linear regression model. Set `n_features_to_select=3`

In [ ]:
# Your code here:


Fit the model and print the ranking

In [ ]:
# Your code here:


Feature importance is ranked from most important (1) to least important (4). Generate a model with the three most important features. The features correspond to variable names. For example, feature 1 is `cylinders` and feature 2 is `displacement`.

Perform a test-train split on this reduced column data and call the split data `X_train_reduced`, `X_test_reduced`, `y_test_reduced`, `y_train_reduced`. Use an 80% split.

In [ ]:
# Your code here:


Generate a new model called `auto_model_reduced` and fit this model. Then proceed to compute the r squared score for the model. Did this cause an improvement in the r squared score?

In [ ]:
# Your code here: 


#### Conclusion

You may obtain the impression from this lab that without knowing statistical methods in depth, it is difficult to make major progress in machine learning. That is correct. If you are motivated to become a data scientist, statistics is the subject you must be proficient in and there is no shortcut. 

Completing these labs is not likely to make you a data scientist. But you will have a good sense about what are there in machine learning and what are good for you. In your future career, you can choose one of the three tracks:

* Data scientists who need to be proficient in statistical methods.

* Data engineers who need to be good at programming.

* Data integration specialists who are business or content experts but also understand data and programming. This cross-disciplinary track brings together data, technology, and business and will be in high demands in the next decade.